agent/consul/rpc.go

```diff
@@ -966,7 +971,15 @@ func (s *Server) blockingQuery(queryOpts structs.QueryOptionsCompat, queryMeta s
 
 		err := fn(ws, state)
 		s.setQueryMeta(queryMeta, queryOpts.GetToken())
-		if err != nil {
+		switch {
+		case errors.Is(err, errNotFound):
+			if notFound {
```

Comment:
Why do we need `var notFound bool`? It seems like it moots the whole point, because the first `errNotFound` skips the conditional here, and proceeds exactly like the prior scenario, never executing line 978.

Reply:
Good question!

I believe what you are calling out is that on the first iteration of this loop we still rely on the index not having changed. If a write happens between the time the last blocking query returned and when the next blocking query request happens (which I think should generally only be a few milliseconds), then this optimization does not work. We continue to have the old behaviour of returning unnecessarily. Generally that should be ok because it only limits the optimization, should not impact correctness, and there is a very small window where a write must happen to cause this problem.

There is a way to work around this limitation (could be a follow up), but it introduces some risk, so I thought it would be better to avoid.  I haven't documented that anywhere yet (just some discussions in old slack threads), but I can do that here.

The work around for this limitation would be to add a new `X-Consul-Index-Not-Found` query parameter (as a sibling to `X-Consul-Index`). This parameter would be used to set the initial value of `var notFound bool`.  That way when a new query comes in, we'll already know if we're in a "not found" state, and only the creation of an item (or a timeout) will cause `blockingQuery` to return.  The risk here is that if  the user incorrectly sets this header, then they could miss a delete that happened, but I haven't thought enough about this case to know if it's actually a problem for correctness. Maybe we should do this as a follow up.

> Why do we need `var notFound bool`?

To answer the other part of this question, this is essential to the optimization.  I think this is explained briefly in 9abb49da51ad87312a5f34364d8e6a3c0e35c328, but I'm happy to add more detail since this is the main reason for the change.  As documented in #12343, there are two things to consider here: the "last modified index" set on `responseMeta.Index`, and the `WatchSet` blocking.

We can't fix the `WatchSet` blocking without turning every read operation into a write that modifies the memdb index (I'm pretty sure this is not an approach we want to investigation).  Maybe there are other ways of making `WatchSet` block correctly for not found items? I have not found any.

So that leaves the index comparison (`responseMeta.Index > query.MinIndex`). Generally we know this comparison will return true in  these "not found" cases, because `responseMeta.Index` will be the max index of the table, and it changes on every write. 

What we need is a comparison we can use instead of `responseMeta.Index > query.MinIndex`.

If we think about this comparison more abstractly, I think we'll see that the question this is asking is "has the result of the query changed from the last result sent to the client". There are plenty of ways that question could be answered, using the modify index is a very cost effective one, so that's generally what we use.  

However as we've noticed, using  that modify index doesn't work for cases where an item does not exist, because we don't have a stable modify index to compare against.  In this case we use  "not found" to indicate if the result has changed. 

Since we don't know the state from the client perspective (`X-Consul-Index-Not-Found` would give us that), we have to perform the query once. If the result of the query is the same "last modified index", then we know at that index the query returned no results. Until the query returns some result, the state has not changed, so we can continue to block until that time.

In the line below this, when we see "state has not changed, the result is still not found", we update the `minQueryIndex` to the latest `responseMeta.Index`, so that the comparison later keeps us in the for loop.

Reply:
I opened #12363 for that potential future optimization.